### PR TITLE
Captioning Service

### PR DESCRIPTION
**Pull request by fearnworks**
First checkpoint for the caption service : 
Includes: 
- Offline batching with vllm (Aria) 
- Guided Decoding with Outlines
- Guided Decoding REST api (pixtral only)
- Several other experiments that will be harvested from 

Grounding is very basic right now and just includes a basic chain of thought with generated tags with default sampling params. A future PR will focus on honing in on optimizing the sampling params to further reduce visual hallucinations.  

Cookbook showing usage of the API can be found here : 
https://github.com/Open-Model-Initiative/OMI-Data-Pipeline/blob/3dced1c5a58db87a21ff89305793d1f3438d136d/modules/odr_caption/cookbook/basic_caption.ipynb

### CODE REVIEW

File: modules/odr_caption/Taskfile.caption.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+version: '3'
+
+# dotenv: ['.env', '{{.ENV}}/.env', '{{.HOME}}/.env']
```
Comment (CheesyLaZanya) [inline modules/odr_caption/Taskfile.caption.yml:3]: Is this comment for future work or should it be removed?

File: modules/odr_caption/odr_caption/utils/message_logger.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# knowledge_worker/graph/message_logger.py
```
Comment (CheesyLaZanya) [inline modules/odr_caption/odr_caption/utils/message_logger.py:1]: Should this comment be removed?

File: modules/odr_caption/odr_caption/server/app.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+"""
+file: main.py
```
Comment (CheesyLaZanya) [inline modules/odr_caption/odr_caption/server/app.py:2]: Should this be app.py?

File: modules/odr_caption/odr_caption/schemas/vllm_schemas.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
+# schemas.py
```
Comment (CheesyLaZanya) [inline modules/odr_caption/odr_caption/schemas/vllm_schemas.py:1]: Should this comment be removed?

File: modules/odr_caption/odr_caption/outlines/caption.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    {
+      "tag": "subject 2",
+      "category": "Entity",
+      "confidence": 0.95
+    },
+    {
+      "tag": "subject 1 runs from subject 2",
+      "category": "Relationship",
+      "confidence": 0.90
+    },
```
Comment (CheesyLaZanya) [inline modules/odr_caption/odr_caption/outlines/caption.py:71]: There is a missing closing ']' bracket in the tags_list example here.

File: modules/odr_caption/odr_caption/outlines/vision.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
+# Load the image
```
Comment (CheesyLaZanya) [inline modules/odr_caption/odr_caption/outlines/vision.py:1]: Should this comment be removed?

File: modules/odr_caption/vllm_deploy/models/vllm_deployment.md [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+     model: my-model-name-or-path
+     tokenizer_mode: auto
+     max_model_len: 8192
+     quantization: None
+     gpu_memory_utilization: 0.9
+     max_num_batched_tokens: 8192
+     max_num_seqs: 256
+     block_size: 16
+     tensor_parallel_size: 1
+     dtype: auto
```
Comment (CheesyLaZanya) [inline modules/odr_caption/vllm_deploy/models/vllm_deployment.md:523]: Missing close ``` ?

File: modules/odr_caption/Taskfile.caption.yml (hunk 1 above)
Comment (fearnworks) [reply modules/odr_caption/Taskfile.caption.yml]: removed

File: modules/odr_caption/odr_caption/outlines/caption.py (hunk 5 above)
Comment (fearnworks) [reply modules/odr_caption/odr_caption/outlines/caption.py:71]: nice catch! Luckily we are not relying on the llm to generate the structure so unlikely to cause an issue, but will fix regardless.

File: modules/odr_caption/docker/Dockerfile.caption [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+# Install system dependencies and Python
+RUN apt-get update && apt-get install -y \
+    python3.11 \
+    python3-pip \
+    python3.11-dev \
+    python3.11-venv \
+    build-essential \
+    exiftool \
+    git \
+    ninja-build \
```
[the remark this answers is not in the document]
Comment (fearnworks) [reply modules/odr_caption/docker/Dockerfile.caption]: resolved

File: modules/odr_caption/vllm_deploy/models/docker/Dockerfile.qwen [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+WORKDIR /vllm-workspace
+
+# Install additional dependencies for Qwen models
+RUN --mount=type=cache,target=/root/.cache/pip \
+    pip install git+https://github.com/huggingface/transformers@21fac7abba2a37fae86106f87fcf9974fd1e3830 \
+    accelerate \
+    hf_transfer \
+    'modelscope!=1.15.0' \
+    bitsandbytes>=0.44.0 \
+    timm==0.9.10
```
[the remark this answers is not in the document]
Comment (fearnworks) [reply modules/odr_caption/vllm_deploy/models/docker/Dockerfile.qwen]: resolved

File: modules/odr_caption/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+openai
+pydantic
+termcolor
+pillow
+python-dotenv
+torch
+torchvision
+httpx
+accelerate
+termcolor
```
Comment (CheesyLaZanya) [inline modules/odr_caption/requirements.txt:15]: termcolor is here twice

File: scripts/setup_virt_env.ps1 [hunk 11]
```diff
@@ -24,5 +24,6 @@ python -m pip install --upgrade pip
 pip install -e ./modules/odr_core
 pip install -e ./modules/odr_api
 pip install -e ./modules/odr_monitoring
+pip install -e ./modules/odr_caption
 pip install -r ./requirements-dev.txt
 pre-commit install
```
Comment (CheesyLaZanya) [inline scripts/setup_virt_env.ps1:29]: Should the shell file also have 'pip install git+https://github.com/dottxt-ai/outlines.git --upgrade' ?

**Review by CheesyLaZanya — APPROVED** [review]
Reviewed, code all looks good.

Wasn't able to test the functionality myself, but OMI agreed best to move this forward for now and just continuing the testing work going forward.

**Pull request merged**